sled-agent/src/rack_setup/plan/service.rs

```diff
@@ -0,0 +1,410 @@
+// This Source Code Form is subject to the terms of the Mozilla Public
+// License, v. 2.0. If a copy of the MPL was not distributed with this
+// file, You can obtain one at https://mozilla.org/MPL/2.0/.
+
+//! Plan generation for "where should services be initialized".
+
+use crate::params::{
+    DatasetEnsureBody, ServiceType, ServiceZoneRequest, ZoneType,
+};
+use crate::rack_setup::config::SetupServiceConfig as Config;
+use omicron_common::address::{
+    get_switch_zone_address, Ipv6Subnet, ReservedRackSubnet, DNS_PORT,
+    DNS_SERVER_PORT, RSS_RESERVED_ADDRESSES, SLED_PREFIX,
+};
+use omicron_common::backoff::{
+    internal_service_policy, retry_notify, BackoffError,
+};
+use serde::{Deserialize, Serialize};
+use sled_agent_client::{
+    types as SledAgentTypes, Client as SledAgentClient, Error as SledAgentError,
+};
+use slog::Logger;
+use std::collections::HashMap;
+use std::net::{Ipv6Addr, SocketAddrV6};
+use std::path::{Path, PathBuf};
+use thiserror::Error;
+use uuid::Uuid;
+
+// The number of Nexus instances to create from RSS.
+const NEXUS_COUNT: usize = 1;
+
+// The number of CRDB instances to create from RSS.
+const CRDB_COUNT: usize = 1;
+
+// TODO(https://github.com/oxidecomputer/omicron/issues/732): Remove
+// when Nexus provisions Oximeter.
+const OXIMETER_COUNT: usize = 1;
+// TODO(https://github.com/oxidecomputer/omicron/issues/732): Remove
+// when Nexus provisions Clickhouse.
+const CLICKHOUSE_COUNT: usize = 1;
+// TODO(https://github.com/oxidecomputer/omicron/issues/732): Remove.
+// when Nexus provisions Crucible.
+const MINIMUM_ZPOOL_COUNT: usize = 3;
+
+fn rss_service_plan_path() -> PathBuf {
+    Path::new(omicron_common::OMICRON_CONFIG_PATH).join("rss-service-plan.toml")
+}
+
+/// Describes errors which may occur while generating a plan for services.
+#[derive(Error, Debug)]
+pub enum PlanError {
+    #[error("I/O error while {message}: {err}")]
+    Io {
+        message: String,
+        #[source]
+        err: std::io::Error,
+    },
+
+    #[error("Cannot deserialize TOML file at {path}: {err}")]
+    Toml { path: PathBuf, err: toml::de::Error },
+
+    #[error("Error making HTTP request to Sled Agent: {0}")]
+    SledApi(#[from] SledAgentError<SledAgentTypes::Error>),
+
+    #[error("Error initializing sled via sled-agent: {0}")]
+    SledInitialization(String),
+
+    #[error("Failed to construct an HTTP client: {0}")]
+    HttpClient(reqwest::Error),
+}
+
+#[derive(Clone, Debug, Default, Deserialize, Serialize, PartialEq)]
+pub struct SledRequest {
+    /// Datasets to be created.
+    #[serde(default, rename = "dataset")]
+    pub datasets: Vec<DatasetEnsureBody>,
+
+    /// Services to be instantiated.
+    #[serde(default, rename = "service")]
+    pub services: Vec<ServiceZoneRequest>,
+
+    /// DNS Services to be instantiated.
+    #[serde(default, rename = "dns_service")]
+    pub dns_services: Vec<ServiceZoneRequest>,
```

Comment:
This is kinda nitpicky and certainly doesn't need to change as part of this PR, but it seems a little strange that this vec has the same type as `services` in the "make illegal states unrepresentable" sense. What happens if there's an item in this vec that doesn't have `ZoneType::InternalDns`, or an item in `services` that does? Would it make sense to split `ServiceZoneRequest` into two types if DNS zones are special enough to be split out here?

Reply:
`ServiceZoneRequest` has an internal `ZoneType` field that allows for matching - it's mostly RSS that cares about the distinction, because it wants to launch DNS services before others.

I'll just lump all services into `pub services`, and make this distinction internally. I agree that it's odd to make illegal state representation possible here.

Done in 8d899c513a13bb809327fa888b70f1cede76e50f